simphony/engine/extension.py

```diff
@@ -0,0 +1,204 @@
+
+
+class EngineManager(object):
+    """Controller class to keep track of engine extensions."""
+    def __init__(self):
+        self._engine_extensions = {}
+
+    def load_metadata(self, plugin):
+        """Load existing engine extensions into the manager.
+
+        This method inspects the given module and looks for ABCEngineExtension
+         subclasses. If provided, will instantiate and keep them for further
+         interactions with corresponding engines.
+
+        plugin: module
+          A python module provided by an extension
+        """
+        if not inspect.ismodule(plugin):
+            raise EngineManagerException(
+                'The provided object is not a module: %s' % plugin)
+
+        for name, value in inspect.getmembers(plugin, inspect.isclass):
+            if not inspect.isabstract(value) and\
+                    issubclass(value, ABCEngineExtension):
+                # Load the corresponding engine extension
+                extension = value()
+                if not isinstance(extension, ABCEngineExtension):
+                    raise ValueError('Expected ABCEngineExtension, got %s' %
+                                     extension)
+                self.add_extension(extension)
+
+    def add_extension(self, extension):
+        """Add an extension to the context.
+
+        Parameters
+        ----------
+        extension: ABCEngineExtension
+          an extension that has knowledge about its own engines.
+        """
+        for engine in extension.get_supported_engines():
+            if engine.name in self._engine_extensions:
+                raise Exception('There is already an extension registered'
+                                ' for %s engine.' % engine.name)
+
+            # This extension knows how to create wrappers for the given engine.
+            # It is our interaction point with its engines.
+            self._engine_extensions[engine.name] = extension
+
+    def create_wrapper(self, cuds, engine_name, engine_interface=None):
+        """Create a wrapper to the given engine.
+
+        Parameters
+        ----------
+        cuds: CUDS
+            A cuds object which contains model information.
+        engine_name: str
+            Name of the underlying engine to launch the simulation with.
+        engine_interface: EngineInterface
+            The interface to the engine, internal or fileio.
+        """
+        if engine_name in self._engine_extensions:
+            extension = self._engine_extensions[engine_name]
+            return extension.create_wrapper(cuds,
+                                            engine_name,
+                                            engine_interface)
+        else:
+            raise EngineManagerException(
+                'Invalid engine name: %s' % engine_name)
+
+    def get_supported_engines(self):
+        """Show a list of supported engines."""
+        return self._engine_extensions.keys()
```

Comment:
This method is intended to be a query function for user to get informed about available engines. I renamed this one to `get_supported_engine_names` and added a new one called `get_supported_engines` which returns a list of EngineMetadata as you mentioned earlier.